File: backend/guardian-absence/src/main/java/com/guardian/absence/domain/Absence.java
package com.guardian.absence.domain;

import java.time.LocalDate;
import java.util.Objects;
import java.util.UUID;

/**
 * A guardian's declaration that a child is not travelling (MOD-14, ABS-001).
 *
 * <p>Its whole purpose is downstream: an active declaration removes the child from the manifest
 * materialised at trip start (BR-ABS-002), which is what stops trip-close reconciliation raising a
 * false alarm about a child who was never coming. An absence is therefore not a note on a profile —
 * it is an input to a safety calculation.
 *
 * @param direction null means <strong>both</strong> journeys. Modelled as a nullable field rather
 *     than a third enum value because that is the wire contract
 *     (guardian-docs/04-api/TRIPS_BOARDING_API.md), and translating between the two shapes at the
 *     boundary is one more place for "both" and "unknown" to be confused.
 * @param reason optional, always. Requiring a parent to justify their child's absence is friction
 *     with no safety value (BR-ABS-001).
 */
public record Absence(
    UUID id,
    UUID studentId,
    UUID declaredByGuardianId,
    LocalDate fromDate,
    LocalDate toDate,
    Direction direction,
    String reason,
    Status status) {

  public Absence {
    Objects.requireNonNull(studentId, "studentId");
    Objects.requireNonNull(fromDate, "fromDate");
    Objects.requireNonNull(toDate, "toDate");
    Objects.requireNonNull(status, "status");

    if (toDate.isBefore(fromDate)) {
      throw new IllegalArgumentException("toDate must not precede fromDate");
    }
  }

  /** Which run the declaration covers. Null on {@link Absence#direction} means both. */
  public enum Direction {
    PICKUP,
    DROP
  }

  /**
   * Cancellation is a status, never a delete.
   *
   * <p>A manifest was materialised from this declaration, so the reason a child was excluded from a
   * trip has to stay answerable after the fact (BR-ABS-004).
   */
  public enum Status {
    ACTIVE,
    CANCELLED
  }

  /** Whether this declaration covers a given date and run. */
  public boolean covers(LocalDate date, Direction run) {
    if (status != Status.ACTIVE) {
      return false;
    }
    boolean inRange = !date.isBefore(fromDate) && !date.isAfter(toDate);
    return inRange && (direction == null || direction == run);
  }
}

File: flutter/parent_app/lib/features/absence/repository/models/absence.dart
import 'package:flutter/foundation.dart';

/// Which leg an absence covers. `null` on the wire means both (BR-ABS-001) — modelled here
/// as an explicit [both] rather than a nullable field, so a screen never has to ask "does
/// null mean both, or unknown?".
enum AbsenceDirection { both, morningOnly, afternoonOnly }

/// A declared absence — P-06's "Aarav will not be expected on Bus 12 tomorrow morning."
@immutable
class Absence {
  const Absence({
    required this.id,
    required this.studentId,
    required this.fromDate,
    required this.toDate,
    required this.direction,
    this.reason,
  });

  final String id;
  final String studentId;

  /// Calendar dates, school-local — not instants (docs/04-api/API_STANDARDS.md: dates are
  /// `YYYY-MM-DD`).
  final DateTime fromDate;
  final DateTime toDate;
  final AbsenceDirection direction;

  /// Optional — requiring a parent to justify an absence is friction with no safety value
  /// (BR-ABS-001).
  final String? reason;
}
